src/CommunityToolkit.Maui/Views/Popup/PopupPage.shared.cs

```diff
@@ -0,0 +1,185 @@
+	readonly WeakEventManager popupClosedEventManager = new();
+
+	public PopupPage(View view, IPopupOptions popupOptions)
+		: this(view as Popup ?? CreatePopupFromView<Popup>(view), popupOptions)
+	{
+		ArgumentNullException.ThrowIfNull(view);
+	}
+
+	public PopupPage(Popup popup, IPopupOptions popupOptions)
+	{
+		ArgumentNullException.ThrowIfNull(popup);
+		ArgumentNullException.ThrowIfNull(popupOptions);
+
+		this.popup = popup;
+		this.popupOptions = popupOptions;
+
+		// Only set the content if parent constructor hasn't set the content already; don't override content if it already exists
+		base.Content ??= new PopupPageLayout(popup, popupOptions);
+
+		tapOutsideOfPopupCommand = new Command(async () =>
+		{
+			popupOptions.OnTappingOutsideOfPopup?.Invoke();
+			await Close(new PopupResult(true));
+		}, () => popupOptions.CanBeDismissedByTappingOutsideOfPopup);
+
+		Content.GestureRecognizers.Add(new TapGestureRecognizer { Command = tapOutsideOfPopupCommand });
+
+		if (popupOptions is BindableObject bindablePopupOptions)
+		{
+			bindablePopupOptions.PropertyChanged += HandlePopupPropertyChanged;
+		}
+
+		this.SetBinding(BindingContextProperty, static (Popup x) => x.BindingContext, source: popup, mode: BindingMode.OneWay);
+		this.SetBinding(BackgroundColorProperty, static (IPopupOptions options) => options.PageOverlayColor, source: popupOptions, mode: BindingMode.OneWay);
+
+		Shell.SetPresentationMode(this, PresentationMode.ModalNotAnimated);
+		On<iOS>().SetModalPresentationStyle(UIModalPresentationStyle.OverFullScreen);
+	}
+
+	public event EventHandler<IPopupResult> PopupClosed
+	{
+		add => popupClosedEventManager.AddEventHandler(value);
+		remove => popupClosedEventManager.RemoveEventHandler(value);
+	}
+
+	// Prevent Content from being set by external class
+	// Casts `PopupPage.Content` to return typeof(PopupPageLayout)
+	internal new PopupPageLayout Content => (PopupPageLayout)base.Content;
+
+	public async Task Close(PopupResult result, CancellationToken token = default)
+	{
+		token.ThrowIfCancellationRequested();
+
+		var popupPageToClose = Navigation.ModalStack.OfType<PopupPage>().LastOrDefault(popupPage => popupPage.Content == Content);
+
+		if (popupPageToClose is null)
+		{
+			throw new InvalidOperationException($"Unable to close popup: could not locate {nameof(PopupPage)}. If using a custom implementation of {nameof(Popup)}, override the {nameof(Close)} method");
+		}
+
+		if (Navigation.ModalStack[^1] is Microsoft.Maui.Controls.Page currentVisibleModalPage
+			&& currentVisibleModalPage != popupPageToClose)
+		{
+			throw new InvalidOperationException($"Unable to close Popup because it is blocked by the Modal Page {currentVisibleModalPage.GetType().FullName}. Please call `{nameof(Navigation)}.{nameof(Navigation.PopModalAsync)}()` to first remove {currentVisibleModalPage.GetType().FullName} from the {nameof(Navigation.ModalStack)}");
+		}
+
+		await Navigation.PopModalAsync(false).WaitAsync(token);
```

Comment:
Do we care to pass the token here? Like, it will just close the modal
```suggestion
		await Navigation.PopModalAsync(false);
```

Reply:
> Do we care to pass the token here 

Yup, we still want to keep it. If a user cancels the `CancellationToken` before we call `await Navigation.PopModalAsync()` it will prevent `Navigation.PopModalAsync()` from running.

Reply:
@TheCodeTraveler I don't think it works this way... If I didn't anything wrong, you can see that the method will be called even if the token is already cancelled.

https://dotnetfiddle.net/fdOP3x

Reply:
I confirmed that `.WaitAsync(CancellationToken)` does indeed throw a `System.Threading.Tasks.TaskCanceledException` when the token is canceled. Here's a slimmed down version of your example that cancels the token immediately instead of waiting a few seconds: https://dotnetfiddle.net/kIU6ij

That said - you did uncover an interesting bug here! I dug into the implementation of `INavigation`, and it turns out that `Navigation.PopModalAsync()` will sometimes return a completed Task. When this happens, the token in `.WaitAsync()` is ignored. Here's the code I'm referring to, found in `NavigationProxy`: https://github.com/dotnet/maui/blob/e5c252ec7f430cbaf28c8a815a249e3270b49844/src/Controls/src/Core/NavigationProxy.cs#L192-L196

```cs
protected virtual Task<Page> OnPopModal(bool animated)
{
	INavigation innerNav = Inner;
	return innerNav is null ? Task.FromResult(PopModal()) : innerNav.PopModalAsync(animated);
}
```

And here's a simple example of `CancellationToken` being ignored in `.WaitAsync()` when a completed Task is returned: https://dotnetfiddle.net/3tfm4W

In other words, you're right - sometimes `.WaitAsync()` will be ignored because `PopupModalAsync()` returns a completed `Task`. 

To fix this, I've now call `token.ThrowIfCancellationRequested();` immediately before we call `await Navigation.PopModalAsync(false).WaitAsync(token);`.

Reply:
Here's the updated code. I added a bunch of comments to help future us. 

https://github.com/CommunityToolkit/Maui/blob/58bda0795b86bec9036d2bbc2b41744b9030ed31/src/CommunityToolkit.Maui/Views/Popup/PopupPage.shared.cs#L75-L104

Reply:
@TheCodeTraveler even with your sample the Task will execute, the thing is not the exception, it will throw, but with the catch behavior you'll see that the `Bla` task is executed no matter what.

https://dotnetfiddle.net/lQjbBD

The cancellation will happen and stop the flow, but the task (in our case the PopAsync) will execute and that's expected to keep the legacy behavior that you can't cancel an ongoing task, more info here

https://devblogs.microsoft.com/dotnet/how-do-i-cancel-non-cancelable-async-operations/

Reply:
I'll keep this discussion solved, since your update makes sure that flow will be correct